src/main/java/io/appform/dropwizard/sharding/observers/bucket/BucketKeyPersistor.java

```diff
@@ -0,0 +1,304 @@
+            return oldSaver.apply(entity);
+        });
+        return null;
+    }
+
+    @Override
+    public <T> Void visit(SaveAll<T> opContext) {
+        val oldSaver = opContext.getSaver();
+        opContext.setSaver((Collection<T> entities) -> {
+            entities.forEach(this::addBucketId);
+            return oldSaver.apply(entities);
+        });
+        return null;
+    }
+
+    @Override
+    public <T> Void visit(CreateOrUpdateByLookupKey<T> createOrUpdateByLookupKey) {
+        val oldMutator = createOrUpdateByLookupKey.getMutator();
+        createOrUpdateByLookupKey.setMutator(result -> {
+            if (result != null) {
+                T value = oldMutator.apply(result);
+                addBucketId(value);
+                return value;
+            }
+            return null;
+        });
+
+        val oldSaver = createOrUpdateByLookupKey.getSaver();
+        createOrUpdateByLookupKey.setSaver((T entity) -> {
+            addBucketId(entity);
+            return oldSaver.apply(entity);
+        });
+        return null;
+    }
+
+    @Override
+    public <T> Void visit(CreateOrUpdate<T> createOrUpdate) {
+        val oldMutator = createOrUpdate.getMutator();
+        createOrUpdate.setMutator(result -> {
+            if (result != null) {
+                T value = oldMutator.apply(result);
+                addBucketId(value);
+                return value;
+            }
+            return null;
+        });
+
+        val oldSaver = createOrUpdate.getSaver();
+        createOrUpdate.setSaver((T entity) -> {
+            addBucketId(entity);
+            return oldSaver.apply(entity);
+        });
+        return null;
+    }
+
+    @Override
+    public <T, U> Void visit(CreateOrUpdateInLockedContext<T, U> createOrUpdateInLockedContext) {
+        val oldMutator = createOrUpdateInLockedContext.getMutator();
+        createOrUpdateInLockedContext.setMutator(entity -> {
+            if (entity != null) {
+                T value = oldMutator.apply(entity);
+                addBucketId(value);
+                return value;
+            }
+            return null;
+        });
+
+        val oldSaver = createOrUpdateInLockedContext.getSaver();
+        createOrUpdateInLockedContext.setSaver((T entity) -> {
+            addBucketId(entity);
+            return oldSaver.apply(entity);
+        });
+        return null;
+    }
+
+    @Override
+    public <T, R> Void visit(Select<T, R> select) {
+        return null;
+    }
+
+    private <T> void addBucketId(T entity) {
+        if (Objects.isNull(entity)) {
+            return;
+        }
+
+        val entitymeta = initialisedEntityMeta.get(entity.getClass().getName());
+        if (Objects.isNull(entitymeta)) {
+            return;
+        }
+
+        val bucketKeyField = entitymeta.getBucketKeyField();
+        val shardingKeyField = entitymeta.getShardingKeyField();
+        if (Objects.isNull(bucketKeyField) || Objects.isNull(shardingKeyField)) {
```

Comment:
When would this condition be hit ? If there is no bucketField or sharding field, ideally entity meta itself should not be present ?